### DISCUSSION TITLE
討論list的各種切法，取出list中指定index並反轉以及對於出現空list的解釋(希望大家可以幫我看看我有沒有想錯

### DISCUSSION THREAD

**Discussion by ghost** [1101-GS4719-C/Python-Community-GS4719-C#74] [Q&A] [answered]
試了一下list切出特定index的語法
上課中有提到
`a = [1,2,3,4,5,6,7,8,9]`
 `a[::-1]`         可以得到反轉的a   =>[9,8,7,6,5,4,3,2,1]

如果想取出[2,3,4,5]並反轉呢?
`a[1:5:-1]`      只會回傳空list

正確的寫法:
`a[1:5:][::-1]`   取出特定index後再對整個list進行反轉(先做`a[1:5:]`取出[2,3,4,5]，再對[2,3,4,5]進行反轉`[2,3,4,5][::-1]`

> 這裡更新一下`a[1:5][::-1]`就可以了，第一個中括號最後的':'是多打的

> 另外，有同學提供新的方法`a[1:5][::-1]和a[4:0:-1]是等效的`

那...第一個寫法為什麼會得到空list呢?
首先
`a[1:8]`
這樣的意思是從index 1 取到 index 7，**沒有指定間隔所以預設為1**

`a[1:8:2]`
這樣的意思是從index 1 取到 index 7，指定取的間隔為2，意味從index 1開始取，要取的index為**前一個取的index+2**，所以是取出a[1],a[3],a[5],a[7] =>取出[2,4,6,8]

所以`a[1:8:-2]`，從index 1 取到 index 7，指定取的間隔為(-2)，意味從index 1開始取，要取的index為**前一個取的index-2**，但是index 1不管加幾次(-2)都不可能取到 index 7，所以會得到空list(至於為甚麼是空list而不是顯示error我想應該python自己的設定，這個我不是很確定，看看有沒有其他人可以幫忙解答?)

再來幾個例子
`a[-1:-5]`     得到空list，因為**沒有指定間隔所以預設為1**，index -1不管加幾次1都不會取到index -4
`a[3::-1]`      第二個引數沒有設定，可能會想說是取index 3~最後一個，但這是在間隔為1時才是這麼解釋，現在間隔為-1，所以取出的應該是a[3],a[2],a[1],a[0]
`a[:5:-2]`     第一個引數沒有設定，直覺是從index 0取到index 4，但因為間隔是(-2)不是1，所以取出的(注意方向順序)應該是a[8],a[6]，**注意沒有a[4]，習慣性地認為a[m:n]是取到index (n-1)也是在沒有輸入指定間隔因此間隔預設1的情況下**，
所以
a[m:n:k]取的範圍用白話解釋的話就是，從index m開始取，(用一個一個取的想法去想)接著要取的index為前一個取的index+k(k可以小於0)，如果剛好取到index=n則停止並且這項不取。如果k大於0，取的index都會小於n；如果k小於0，取的index都會大於n

~~另外，>=0的index和負0的index不能一起用，在負的index中沒有0，~~ ` a[-1:5]`這樣出來的也是空list

> 10.29更新，感謝同學回覆說負的index和正的index可以並用!但是要符合前面提到邏輯。對電腦來說`a[-1]`還是`a[8]`都是一樣的「物件」(應該是吧，錯誤的話麻煩糾正我謝謝)，所以看到`a[5:-1]`，可以想成就是`a[5:8:1]`。而` a[-1:5]`是空list的原因是將其看成`a[8:5:1]`違反了前面提到的邏輯
想看更多測試內容[請點這](https://github.com/1101-GS4719-C/Python-Community-GS4719-C/discussions/74#discussioncomment-1557746)

以上是我自己測試的結果，可能有些解釋錯誤/寫不清楚的地方，還請大家多多指教

**Comment by Joan0518** [comment]
感謝分享~~~我認為講解得很清楚><(雖然我還是理解了很久哈哈)

**Comment by madeuf** [comment]
很清楚喔，謝謝同學的分享!!(已收藏

**Comment by chia-kai-chang** [comment]
寫的很棒，讚讚

**Comment by 109802503gloria** [comment, marked as answer]
我之前自己試過，如果想要取出特定index同時對整個list進行反轉，除了同學寫的`a[1:5:][::-1]`，也可以寫成`a[4:0:-1]`，這樣也沒有違反同學的邏輯，只是切割的起始點和結束點稍微需要想一下。

**Comment by enoch-lightning** [comment]
我也自己去測試，發現非負的索引(index)可以與負的索引一起使用。以下舉三個實例。
當`a=[1, 2, 3, 4, 5, 6, 7, 8, 9]`
若取

1. `a[0:-1]`會得到`[1, 2, 3, 4, 5, 6, 7, 8]`
![image](https://user-images.githubusercontent.com/86505454/139367508-a8728f9d-6a97-4d0e-a120-ef1e3a3d3281.png)
2. `a[-1:0:-1]`會得到`[9, 8, 7, 6, 5, 4, 3, 2]
![image](https://user-images.githubusercontent.com/86505454/139368218-0a770152-7f69-4aeb-b748-f2bab8461b5e.png)
3. `a[5:-1]`會得到`[6, 7, 8]`
<img width="960" alt="未命名" src="https://user-images.githubusercontent.com/86505454/139370072-efd9af97-8142-4a6f-87c6-c49d58313bff.png">

**Reply by ghost** [reply]
感謝 @enoch-lightning 的糾正，我會再更新到內文中

我剛剛試了一下，簡單來說，想法一樣不變，只要把負的index換回正的index就一樣可以用文內提到的邏輯去想，對電腦來說a[-1]還是a[8]都是一樣的東西(一樣的值指向相同的記憶體位置。...是不是應該講物件?XD)，看下圖中的id就知道了。有問題的是如果不轉回同號的index，我們人在想邏輯的時候就會卡卡的
![image](https://user-images.githubusercontent.com/90567166/139407681-229e1931-11ba-48a7-8d53-d635589e8707.png)

注意: 雖然a[-1] 、 a[8]的id相同，但是id(a[0:-1])和id(a[0:8])會不同，我試了一下後覺得是因為list是mutable物件，所以就算是長的一模一樣的list，為了要有mutable的特性，所以不能指向相同的記憶體位址(白話一點，就是看下圖，因為y要能append之後仍指向同一個記憶體位置，為了保有這個特性，所以即使原本還沒append的y和x是相等的，也不能指向同一個記憶體位置，不然y.append之後，如果x是指向同一個記憶體位置，那x就要跟著改變了，這是錯的!)

![2](https://user-images.githubusercontent.com/90567166/139414604-ab35cd51-8490-47f1-8111-add2303a114a.jpg)
